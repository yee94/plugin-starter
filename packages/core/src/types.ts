export interface Plugin {
  /**
   * The name of the plugin.
   */
  name: string;

  /**
   * The enforced order of the plugin.
   */
  enforce?: 'pre' | 'post';

  /**
   * Modify config before it's resolved. The hook can either mutate the
   */
  config?: (this: void, config: UserConfig) => UserConfig | null | void | Promise<UserConfig | null | void>;
}

export type PluginFunctionKeys<T = Required<Plugin>> = {
  [K in keyof T]: T[K] extends Function ? K : never;
}[keyof T];


export type PluginOption =
  | Plugin
  | false
  | null
  | undefined
  | PluginOption[]
  | Promise<Plugin | false | null | undefined | PluginOption[]>;
export interface UserConfig {
  /**
   * Array of plugins to use.
   */
  plugins?: PluginOption[];
}

export const defineConfig = (config: UserConfig) => config;
