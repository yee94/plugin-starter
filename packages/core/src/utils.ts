import mergeWith from 'lodash/mergeWith';

export async function asyncFlatten<T>(arr: T[]): Promise<T[]> {
  do {
    arr = (await Promise.all(arr)).flat(Infinity) as any;
  } while (arr.some((v: any) => v?.then));
  return arr;
}

export function deepMerge(a, b) {
  return mergeWith(a, b, (objValue: any, srcValue: any) => {
    if (Array.isArray(objValue)) {
      return objValue.concat(srcValue);
    }
  });
}
