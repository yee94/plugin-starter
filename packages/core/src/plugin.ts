import { Plugin, PluginFunctionKeys } from './types';
import { deepMerge } from './utils';

export function createPluginContainer(plugins: Plugin[]) {
  return async <T extends PluginFunctionKeys>(
    fnName: T,
    args: Parameters<NonNullable<Plugin[T]>>,
    options?: {
      iterate?: <R = ReturnType<NonNullable<Plugin[T]>> | null>(prev: R, next: R) => R;
    },
  ) => {
    const iterate = options?.iterate ?? deepMerge;
    let prev: any = args[0];
    for (const p of plugins) {
      const handler = p[fnName];
      if (handler) {
        const next = await handler.apply(void 0, args);
        if (next) {
          args[0] = next;
        }

        prev = iterate(prev, next);
      }
    }

    return prev;
  };
}
