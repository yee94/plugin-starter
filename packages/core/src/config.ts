import { Plugin, UserConfig } from './types';
import { asyncFlatten, deepMerge } from './utils';
import { createPluginContainer } from './plugin';

/**
 * Resolve the config into a usable format.
 * @param config
 */
export async function resolveConfig(config: UserConfig) {
  const rawWorkerUserPlugins = ((await asyncFlatten(config.plugins || [])) as Plugin[]).filter(Boolean);

  const [workerPrePlugins, workerNormalPlugins, workerPostPlugins] = sortUserPlugins(rawWorkerUserPlugins);

  let workerConfig = config;

  const workerUserPlugins = [...workerPrePlugins, ...workerNormalPlugins, ...workerPostPlugins];

  const pluginContainer = createPluginContainer(workerUserPlugins);
  workerConfig = await pluginContainer('config', [workerConfig], {
    iterate: mergeConfig as any,
  });

  workerConfig.plugins = workerUserPlugins;

  return workerConfig;
}

async function runConfigHook(config: UserConfig, plugins: Plugin[]): Promise<UserConfig> {
  let conf = config;

  for (const p of plugins) {
    const handler = p.config;
    if (handler) {
      const res = await handler(conf);
      if (res) {
        conf = mergeConfig(conf, res);
      }
    }
  }

  return conf;
}

/**
 * Merge two configs together.
 * @param defaults
 * @param overrides
 */
export function mergeConfig(defaults: Partial<UserConfig>, overrides: Partial<UserConfig>): UserConfig {
  return deepMerge(defaults, overrides);
}

export function sortUserPlugins(plugins: (Plugin | Plugin[])[] | undefined): [Plugin[], Plugin[], Plugin[]] {
  const prePlugins: Plugin[] = [];
  const postPlugins: Plugin[] = [];
  const normalPlugins: Plugin[] = [];

  if (plugins) {
    plugins.flat().forEach((p) => {
      if (p.enforce === 'pre') prePlugins.push(p);
      else if (p.enforce === 'post') postPlugins.push(p);
      else normalPlugins.push(p);
    });
  }

  return [prePlugins, normalPlugins, postPlugins];
}
