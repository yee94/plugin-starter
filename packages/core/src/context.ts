import { createPluginContainer } from './plugin';

export class Context {
  // @ts-ignore
  pluginContainer: ReturnType<typeof createPluginContainer>;
}
