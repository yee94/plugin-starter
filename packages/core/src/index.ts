import { Context } from './context';
import { resolveConfig } from './config';
import { createPluginContainer } from './plugin';
import { UserConfig } from './types';

export function createInstance() {
  return {
    ctx: new Context(),
    async init() {
      const config: UserConfig = {}; // Load from file, can use resolveConfig from esbuild-resolve-config

      const resolvedConfig = await resolveConfig(config);

      this.ctx.pluginContainer = createPluginContainer(resolvedConfig.plugins);
    },
  };
}
