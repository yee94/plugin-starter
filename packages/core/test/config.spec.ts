import { resolveConfig } from '../src/config';
import { Plugin, PluginOption } from '../src/types';
import {expect} from "vitest"

describe('config', () => {
  it('should be get config from plugin', async () => {
    const result = await resolveConfig({
      plugins: [
        {
          name: 'test',
          config: async (config) => {
            return {
              ...config,
              test: {
                globals: true,
                passWithNoTests: true,
              },
            };
          },
        },
      ],
    });

    expect(result).toMatchInlineSnapshot(`
      {
        "plugins": [
          {
            "config": [Function],
            "name": "test",
          },
        ],
        "test": {
          "globals": true,
          "passWithNoTests": true,
        },
      }
    `);
  });

  it('should be flat plugin array', async () => {
    const preset = () =>
      [
        [
          {
            name: 'test',
          },
          null,
          {
            name: 'test2',
          },
          {
            name: 'test3',
            enforce: 'pre',
          },
        ],
        {
          name: 'test4',
        },
      ] as PluginOption;

    const result = await resolveConfig({
      plugins: [
        preset(),
        {
          name: 'test5',
          enforce: 'pre',
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  it('configResolved should be called', async () => {
    const fn = vi.fn();

    const result = await resolveConfig({
      plugins: [
        {
          name: 'test5',
          configResolved: fn,
        },
      ],
    });

    expect(fn).toBeCalled();
    expect(result). toMatchInlineSnapshot(`
      {
        "plugins": [
          {
            "configResolved": [MockFunction spy] {
              "calls": [
                [
                  [Circular],
                ],
              ],
              "results": [
                {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "name": "test5",
          },
        ],
      }
    `)
  });
});
