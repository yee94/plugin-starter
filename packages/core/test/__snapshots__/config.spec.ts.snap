// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > should be flat plugin array 1`] = `
{
  "plugins": [
    {
      "enforce": "pre",
      "name": "test3",
    },
    {
      "enforce": "pre",
      "name": "test5",
    },
    {
      "name": "test",
    },
    {
      "name": "test2",
    },
    {
      "name": "test4",
    },
  ],
}
`;
